src/ast/parse.rs

```diff
@@ -94,801 +92,785 @@ impl Display for Span {
-    let (_, span, next) = ident(ctx)?;
-    match span.string() {
-        "NaN" => Ok((OrderedFloat(f64::NAN), span, next)),
-        "inf" => Ok((OrderedFloat(f64::INFINITY), span, next)),
-        "-inf" => Ok((OrderedFloat(f64::NEG_INFINITY), span, next)),
-        _ => match span.string().parse::<f64>() {
-            Err(_) => Err(ParseError::Float(span)),
-            // Rust will parse "infinity" as a float, which we don't want
-            // we're only using `parse` to avoid implementing it ourselves anyway
-            Ok(x) => match x.classify() {
-                Nan | Infinite => Err(ParseError::Float(span)),
-                Zero | Subnormal | Normal => Ok((OrderedFloat(x), span, next)),
-            },
-        },
+    fn advance_char(&mut self) {
+        assert!(self.index < self.source.contents.len());
+        loop {
+            self.index += 1;
+            if self.source.contents.is_char_boundary(self.index) {
+                break;
+            }
+        }
+    }
+
+    fn advance_past_whitespace(&mut self) {
+        let mut in_comment = false;
+        loop {
+            match self.current_char() {
+                None => break,
+                Some(';') => in_comment = true,
+                Some('\n') => in_comment = false,
+                Some(c) if c.is_whitespace() => {}
+                Some(_) if in_comment => {}
+                Some(_) => break,
+            }
+            self.advance_char();
+        }
     }
-}
 
-fn string(ctx: &Context) -> Res<String> {
-    let mut span = Span(ctx.source.clone(), ctx.index, ctx.index);
-    if ctx.current_char() != Some('"') {
-        return Err(ParseError::String(span));
+    fn is_at_end(&self) -> bool {
+        self.index == self.source.contents.len()
     }
 
-    let mut next = ctx.clone();
-    let mut in_escape = false;
+    fn next(&mut self) -> Result<(Token, Span), ParseError> {
+        self.advance_past_whitespace();
+        let mut span = Span(self.source.clone(), self.index, self.index);
 
-    next.advance_char();
-    loop {
-        match next.current_char() {
-            None => {
-                span.2 = next.index;
-                return Err(ParseError::MissingEndQuote(span));
+        let Some(c) = self.current_char() else {
+            return error!(span, "unexpected end of file");
+        };
+        self.advance_char();
+
+        let token = match c {
+            '(' => Token::Open,
+            ')' => Token::Close,
+            '"' => {
+                let mut in_escape = false;
+                let mut string = String::new();
+
+                loop {
+                    match self.current_char() {
+                        None => {
+                            span.2 = self.index;
+                            return error!(span, "string is missing end quote");
+                        }
+                        Some('"') if !in_escape => break,
+                        Some('\\') if !in_escape => in_escape = true,
```

Comment:
Nit: This means the only escaped characters are `\"` and `\\`. Maybe we can reject other escaped characters like `\n`, `\t`? This can help prevent user's misunderstanding.